LPS-83990 Portlet 3.0: Develop bean portlet Blade template and Maven archetype

ngriffin7a opened the pull request:
@gamerson Here is the bean portlet project template for your review. I think I got it pretty much correct but please double check, especially modules/sdk/project-templates/project-templates-cdi-bean-portlet/src/main/resources/archetype-resources/build.gradle.

I was able to generate a project using the following command:
```
mvn archetype:generate \
    -DarchetypeGroupId=com.liferay \
    -DarchetypeArtifactId=com.liferay.project.templates.cdi.bean.portlet \
    -DarchetypeVersion=1.0.0 \
    -DgroupId=com.mycompany \
    -DartifactId=com.mycompany.foo.bean.portlet
```
Note that when a project is generated from this template, it will contain the following two dependencies:
```
<dependency>
	<groupId>com.liferay</groupId>
	<artifactId>com.liferay.bean.portlet.api</artifactId>
	<version>1.0.0</version>
</dependency>
<dependency>
	<groupId>com.liferay</groupId>
	<artifactId>com.liferay.bean.portlet.cdi.extension</artifactId>
	<version>1.0.0</version>
</dependency>
```
My goal is to submit these dependencies for review to @shuyangzhou this week. The `groupId` and `artifactId` values are subject to change by Brian.

/cc @jhinkey @sez11a @vsingleton

liferay-continuous-integration commented:
To conserve resources, the PR Tester does not automatically run for every pull.

If your code changes were already tested in another pull, reference that pull in this pull so the test results can be analyzed.

If your pull was never tested, comment &quot;ci:test&quot; to run the PR Tester for this pull.

ngriffin7a commented:
/cc @rotty3000

gamerson commented:
hey @ngriffin7a  this is good stuff! Thanks for sending this, I will review it and give detailed feedback tomorrow.  Quick question... do we want to show some view technology that a portlet 3/CDI developer would be very familiar with?  JSPs or JSF?

ngriffin7a commented:
@gamerson JSP perhaps but not JSF since [JSR 378](https://www.jcp.org/en/jsr/detail?id=378) won't be complete until sometime in 2019.

Should the JSP be contained in a separate template? I figured I would keep this one simple because it is a copy of the plain-and-simple com.liferay.project.templates.portlet which does not have a JSP.

gamerson commented:
If you wanted to make it a little more complete we could put in a simple
jsp, but I don't feel too strongly either way.  But since this one does not
implement MVC portlet, I'm not for sure if there is anything special that a
user would need to do to dispatch to a view of some kind.  If it is just a
simple thing then I guess I wouldn't bother.

On Mon, Jul 30, 2018 at 5:10 PM, Neil Griffin <notifications@github.com>
wrote:

> @gamerson <https://github.com/gamerson> JSP perhaps but not JSF since JSR
> 378 <https://www.jcp.org/en/jsr/detail?id=378> won't be complete until
> sometime in 2019.
>
> Should the JSP be contained in a separate template? I figured I would keep
> this one simple because it is a copy of the plain-and-simple
> com.liferay.project.templates.portlet which does not have a JSP.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gamerson/liferay-portal/pull/170#issuecomment-409028606>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkVFVuC_SBQ5v8uZe5vDflQ_7gylBCWks5uL4RpgaJpZM4VnPDG>
> .
>



-- 
Greg Amerson
Liferay Backend Developer Tools
Liferay, Inc. www.liferay.com

ngriffin7a commented:
@gamerson I will update the branch today with another `@RenderMethod` that auto-dispatches to a JSP. I'll let you know when I push the commits so that you can review.

gamerson commented on modules/sdk/project-templates/project-templates-cdi-bean-portlet/src/main/resources/archetype-resources/build.gradle:
```diff
@@ -0,0 +1,43 @@
+	}
+}
+
+apply plugin: "com.liferay.plugin"
```
Hey @ngriffin7a  take a look at the other war projects and see how their build.gradle are configured and see that they must apply the war plugin on the gradle side.  

also, have you been able to test this project template from both gradle and maven and compared the WAR files to make sure they are virtually identical?

ngriffin7a commented:
@gamerson I incorporated all the changes/suggestions and fixed the build.gradle -- verified that the WAR built from a generated project is the same with both Maven and Gradle.

I squashed it all into a single commit. Please review and send it to Brian if you think it is ready. Thanks.

ngriffin7a commented:
@gamerson Do you have time in your schedule to review this today or Monday? Thanks, Neil.

gamerson commented:
Hey Neil,

Will this portlet template work on master?

On Fri, Aug 3, 2018 at 11:04 AM, Neil Griffin <notifications@github.com>
wrote:

> @gamerson <https://github.com/gamerson> Do you have time in your schedule
> to review this today or Monday? Thanks, Neil.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gamerson/liferay-portal/pull/170#issuecomment-410299942>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkVFfbLJNUWeHNuZ69Ye0ALGzpeRkujks5uNHSkgaJpZM4VnPDG>
> .
>



-- 
Greg Amerson
Liferay Backend Developer Tools
Liferay, Inc. www.liferay.com

ngriffin7a commented:
@gamerson It will work in master with the exception of the com.liferay.bean.portlet.cdi.extension dependency which is currently under review https://github.com/shuyangzhou/liferay-portal/pull/6576

Note that it will require follow-up PRs throughout the week, adding implementation in small reviewable chunks. My best guess is that it will be published to repository.liferay.com by Friday or early next week.

gamerson commented on modules/sdk/project-templates/project-templates-cdi-bean-portlet/bnd.bnd:
```diff
@@ -0,0 +1,9 @@
+Bundle-Description: Creates a Liferay CDI bean portlet as a WAR project.
```
hey @ngriffin7a   the name of this template should be `project-templates-war-cdi-bean-portlet` because all of our `war` projects have `war` in the name.

gamerson commented on modules/sdk/project-templates/project-templates-cdi-bean-portlet/src/main/resources/archetype-resources/build.gradle:
```diff
@@ -0,0 +1,31 @@
+	compileOnly group: "javax.portlet", name: "portlet-api"
+	compileOnly group: "javax.servlet", name: "javax.servlet-api"
+	compileOnly group: "jstl", name: "jstl"
+	compile group: "org.jboss.weld.servlet", name: "weld-servlet"
```
Lets group the `compile` all together.

gamerson commented on modules/sdk/project-templates/project-templates-cdi-bean-portlet/src/main/resources/archetype-resources/build.gradle:
```diff
@@ -0,0 +1,31 @@
+	maven {
+		url "https://repository-cdn.liferay.com/nexus/content/groups/public"
+	}
+}#end
```
If you expect people to add css to these portlets, we should likely include the section on configuring the war plugin.

```  See this project template file: https://github.com/liferay/liferay-portal/blob/master/modules/sdk/project-templates/project-templates-war-mvc-portlet/src/main/resources/archetype-resources/build.gradle#L39-L48
war {
	dependsOn buildCSS
	exclude "**/*.scss"

	filesMatching("**/.sass-cache/") {
		it.path = it.path.replace(".sass-cache/", "")
	}

	includeEmptyDirs = false
}
```

gamerson commented on modules/sdk/project-templates/project-templates-cdi-bean-portlet/src/main/resources/archetype-resources/pom.xml:
```diff
@@ -0,0 +1,84 @@
+				</configuration>
+			</plugin>
+		</plugins>
+	</build>
```
If this project is likely going to have CSS, we could add this to the pom

https://github.com/liferay/liferay-portal/blob/master/modules/sdk/project-templates/project-templates-war-mvc-portlet/src/main/resources/archetype-resources/pom.xml#L65-L103

gamerson commented on modules/sdk/project-templates/project-templates-cdi-bean-portlet/src/main/resources/archetype-resources/src/main/java/portlet/__className__Portlet.java:
```diff
@@ -0,0 +1,39 @@
+		include = "/WEB-INF/jsp/view.jsp",
+		portletNames = {${className}PortletKeys.${className}})
+	public String doView() {
+		return "Hello from " + portletConfig.getPortletName();
```
What does the "doView()" method do if there is a JSP?  as include?  before or after?

gamerson replied:
Thanks a sweet annotation btw!

ngriffin7a replied:
To answer your question, the portlet will render the following:
```<p>Hello from Foo</p>```
```<p>Hello from view.jsp</p>```

ngriffin7a commented:
See https://github.com/gamerson/liferay-portal/pull/185
